geth/jail/internal/loop/loop.go

```diff
@@ -51,9 +54,19 @@ func New(vm *vm.VM) *Loop {
 // queue, the capacity of which being specified by the backlog argument.
 func NewWithBacklog(vm *vm.VM, backlog int) *Loop {
 	return &Loop{
-		vm:    vm,
-		tasks: make(map[int64]Task),
-		ready: make(chan Task, backlog),
+		vm:        vm,
+		tasks:     make(map[int64]Task),
+		ready:     make(chan Task, backlog),
+		accepting: true,
+	}
+}
+
+func (l *Loop) close() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if l.accepting {
+		l.accepting = false
```

Comment:
All operations with `l.accepting` isn't thread safe.

Reply:
I'm not sure if I understand.
- The thread acquires the lock, and no other thread will have it.
- Check to see if the loop is accepting tasks
- Perform the appropriate operation

I thought this would be thread-safe.

Reply:
@JekaMas Could you mind elaborating abit on the thread-safeness you mentioned earlier here?

Reply:
My bad, it was late night when i wrote that. This is thread safe but dangerous: we need to always remember that `l.accepting` should be guarded with mutes. Any mistake and we'll get either data race or deadlock.
I think i could be better to extract `accepting` as new type `state`, let him setters, getters that do all mutex incapsulation. `Loop` is in `has-a` relation with `state` so the last one could be embeded into Loop and hide all mutex logic.

Reply:
Ok. I can work on this as well. Thanks for your input. I am really enjoying this PR.

Reply:
I sat down to finally work on this today and I am having an issue. If I am going to extract the mutex logic for `accepting`, then I feel it would be best to extract all mutex logic for `loop` because `tasks` and `ready` also need to be protected by the mutex. However, then the `state` type would just become the `loop` type.

Reply:
@thedadams I believe @JekaMas was trying to imply this:
```go
type state struct {
    b bool
    mu sync.RWMutex
}

type Loop struct {
    ...
    state state
    lock sync.RWMutex
}
```
Thus:
```go
func (l *Loop) Add(t Task) error {
    if !l.state.get() {  // or with a wrapper such as l.accepting()
        // return error
    }
    l.Lock()
    // add the task
    l.Unlock()
    return nil
}
```